File: scripts/convert-to-copilot.py
#!/usr/bin/env python3
"""
convert_to_copilot.py - Thin wrapper for Copilot conversion

This is a Python replacement for convert-to-copilot.js that delegates to
the CopilotConverter in sync-skills.py.

IMPORTANT: This wrapper only creates the custom-instructions format (1 of 3).
The JS converter creates 3 formats:
  1. custom-instructions/{category}/{slug}.md        - ✓ Created by this wrapper
  2. agent-skills/{category}/{slug}/SKILL.md         - ✗ Not created (JS only)
  3. instructions/{category}-{slug}.instructions.md  - ✗ Not created (JS only)

To create all 3 formats, use the original JS converter:
    node convert-to-copilot.js

Usage:
    python convert_to_copilot.py                    # Convert all skills
    python convert_to_copilot.py --dry-run          # Preview changes
    python convert_to_copilot.py --validate         # Run validation
    python convert_to_copilot.py --stats            # Show statistics
    python convert_to_copilot.py --skill technical/api-development
    python convert_to_copilot.py --category ai-agents

This wrapper uses the shared Python infrastructure in sync-skills.py,
enabling integration with lib/ modules (platform_tuning, metadata_enricher,
skill_validator) that the JS converter cannot access.
"""

import argparse
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))


def main():
    """Parse arguments and delegate to sync-skills.py with copilot target."""
    parser = argparse.ArgumentParser(
        description="Convert master skills to GitHub Copilot format (wrapper for sync-skills.py)",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
examples:
  python convert_to_copilot.py
  python convert_to_copilot.py --dry-run
  python convert_to_copilot.py --validate --stats
  python convert_to_copilot.py --skill technical/api-development
  python convert_to_copilot.py --category ai-agents

This script is a thin wrapper that calls:
  python sync-skills.py --targets copilot [options]
        """,
    )

    parser.add_argument(
        "--skill",
        help="Convert a specific skill (e.g., technical/api-development)",
    )
    parser.add_argument(
        "--category",
        help="Convert a specific category (e.g., ai-agents)",
    )
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Run skill validation before converting",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Show what would be converted without writing files",
    )
    parser.add_argument(
        "--stats",
        action="store_true",
        help="Show conversion statistics",
    )

    args = parser.parse_args()

    # Import and use sync-skills.py's infrastructure
    try:
        # Add project root to path to import sync-skills module
        current_dir = Path(__file__).parent.parent.resolve()
        sys.path.insert(0, str(current_dir))

        # Import the syncer (sync-skills module)
        # We can't import it directly as "sync-skills" due to the hyphen,
        # so we'll use importlib
        import importlib.util
        sync_skills_path = current_dir / "scripts" / "sync-skills.py"

        if not sync_skills_path.exists():
            print(f"Error: sync-skills.py not found at {sync_skills_path}", file=sys.stderr)
            print("This wrapper requires sync-skills.py to be in the scripts/ directory.", file=sys.stderr)
            sys.exit(1)

        spec = importlib.util.spec_from_file_location("sync_skills", sync_skills_path)
        sync_skills = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(sync_skills)

    except Exception as e:
        print(f"Error loading sync-skills.py: {e}", file=sys.stderr)
        sys.exit(1)

    # Create syncer instance
    syncer = sync_skills.SkillsSyncer(current_dir)
    targets = ['copilot']

    # Print header
    print("=" * 60)
    print("  GitHub Copilot Skill Converter (Python)")
    print("=" * 60)
    print("  Using: sync-skills.py CopilotConverter")
    print(f"  Output: {current_dir / 'platforms' / 'github-copilot' / 'copilot-skills'}")
    print(f"  Format: custom-instructions/{'{category}'}/{'{slug}'}.md")
    print()
    print("  Note: This creates only custom-instructions format.")
    print("        For agent-skills and path-specific instructions,")
    print("        use: node convert-to-copilot.js")
    print()

    # Common kwargs for sync methods
    common_kwargs = {
        'validate': args.validate,
        'dry_run': args.dry_run,
        'show_stats': args.stats,
    }

    # Dispatch to appropriate sync method
    try:
        if args.skill:
            syncer.sync_skill(args.skill, targets, **common_kwargs)
        elif args.category:
            syncer.sync_category(args.category, targets, **common_kwargs)
        else:
            syncer.sync_all(targets, **common_kwargs)
    except Exception as e:
        print(f"Error during conversion: {e}", file=sys.stderr)
        import traceback
        traceback.print_exc()
        sys.exit(1)

    print()
    print("  Conversion complete.")
    print()


if __name__ == "__main__":
    main()

File: scripts/convert-to-copilot.js
/**
 * @deprecated This converter is deprecated. Use the Python equivalent instead:
 *   python sync-skills.py --targets copilot
 *
 * This JS converter duplicates CopilotConverter in sync-skills.py and cannot
 * share lib/ modules. It will be removed in a future version.
 */

/**
 * convert-to-copilot.js
 *
 * Converts master skill definitions into multiple GitHub Copilot-specific formats:
 *   1. custom-instructions/{category}/{slug}.md        - Enhanced instruction files
 *   2. agent-skills/{category}/{slug}/SKILL.md         - Agent skill format
 *   3. instructions/{category}-{slug}.instructions.md  - Path-specific instructions
 *   4. Category README files + root copilot-instructions-root.md
 *
 * Source: _master-skills/{category}/{skill-name}/SKILL.md
 * Output: platforms/github-copilot/copilot-skills/
 *
 * Node.js only -- no external dependencies.
 */

const fs = require('fs');
const path = require('path');

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const SOURCE_BASE = path.join(__dirname, '..', '_master-skills');
const TARGET_BASE = path.join(__dirname, '..', 'platforms', 'github-copilot', 'copilot-skills');

const CATEGORIES = ['technical', 'strategy', 'creative', 'industry', 'operations', 'ai-agents'];

/** Glob patterns that Copilot path-specific instructions use per category. */
const APPLY_TO_PATTERNS = {
    technical:  '**/*.{ts,js,py,go,rs,java}',
    strategy:   '**/*.{md,txt}',
    creative:   '**/*.{md,html,css,svg}',
    operations: '**/*.{yml,yaml,json,toml,sh}',
    industry:   '**/*.{md,txt,csv,json}',
    'ai-agents': '**/*.{md,yml,yaml,json,py,ts}',
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** Convert a slug like "api-design-principles" to "Api Design Principles". */
function toTitleCase(str) {
    return str
        .split('-')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
}

/** Ensure a directory exists, creating it recursively if needed. */
function ensureDir(dirPath) {
    if (!fs.existsSync(dirPath)) {
        fs.mkdirSync(dirPath, { recursive: true });
    }
}

/**
 * Parse a SKILL.md file. Returns { name, description, body }.
 * Handles files with or without YAML-ish frontmatter delimited by ---.
 */
function parseSkillMd(content) {
    const match = content.match(/^---\r?\n([\s\S]*?)\r?\n---\r?\n([\s\S]*)$/);
    if (!match) {
        return { name: '', description: '', body: content.trim() };
    }

    const frontmatter = match[1];
    const body = match[2].trim();

    const nameMatch = frontmatter.match(/^name:\s*(.+)$/m);
    const descMatch = frontmatter.match(/^description:\s*(.+)$/m);

    return {
        name: nameMatch ? nameMatch[1].trim() : '',
        description: descMatch ? descMatch[1].trim() : '',
        body,
    };
}

/**
 * Return the first N non-empty lines from a body string.
 * Used for the truncated instructions format.
 */
function firstNLines(text, n) {
    const lines = text.split(/\r?\n/);
    const collected = [];
    for (const line of lines) {
        collected.push(line);
        if (collected.length >= n) break;
    }
    return collected.join('\n');
}

// ---------------------------------------------------------------------------
// Format generators
// ---------------------------------------------------------------------------

/**
 * 1. Enhanced custom-instructions format.
 */
function generateCustomInstruction(displayName, description, body, category, slug) {
    const applyTo = APPLY_TO_PATTERNS[category] || '**/*';

    return `# ${displayName}

## Overview
${description || 'No description provided.'}

## Workspace Context
When working with this skill in a codebase:
- Reference relevant files using @workspace
- Use #file:path syntax to include specific file context
- Consider the project's existing patterns and conventions

## Instructions
${body}

## File Patterns
Applies to: \`${applyTo}\`

## Integration Notes
- Use \`@workspace\` to reference project structure
- Use \`@terminal\` for command execution context
- Combine with other instructions in \`.github/copilot-instructions.md\`
`;
}

/**
 * 2. Agent-skills SKILL.md format (with frontmatter).
 */
function generateAgentSkill(name, description, body) {
    return `---
name: ${name}
description: ${description || 'No description provided.'}
---

${body}
`;
}

/**
 * 3. Path-specific .instructions.md format.
 */
function generatePathInstruction(body, category) {
    const applyTo = APPLY_TO_PATTERNS[category] || '**/*';
    const truncatedBody = firstNLines(body, 50);

    return `---
applyTo: "${applyTo}"
---

${truncatedBody}
`;
}

// ---------------------------------------------------------------------------
// Category README
// ---------------------------------------------------------------------------

function generateCategoryReadme(category, skills) {
    const title = toTitleCase(category);
    const skillList = skills
        .map(s => `- [${s.displayName}](./custom-instructions/${category}/${s.slug}.md)`)
        .join('\n');

    return `# ${title} -- Copilot Skills

This directory contains GitHub Copilot skill files for the **${title}** category.

## Available Skills (${skills.length})

${skillList}

## Formats

| Format | Path | Purpose |
|--------|------|---------|
| Custom Instructions | \`custom-instructions/${category}/\` | Full instruction files for \`.github/copilot-instructions.md\` |
| Agent Skills | \`agent-skills/${category}/\` | Skill definitions for Copilot agent mode |
| Path Instructions | \`instructions/\` | File-pattern-scoped instruction snippets |

## Usage

1. **Single Skill** -- Copy content from \`custom-instructions/${category}/<skill>.md\` into your project's \`.github/copilot-instructions.md\`.
2. **Multiple Skills** -- Concatenate the Instructions sections from several files.
3. **Agent Mode** -- Place \`agent-skills/${category}/<skill>/SKILL.md\` where your Copilot agent config expects skill definitions.
4. **Path-Specific** -- Drop \`instructions/${category}-<skill>.instructions.md\` into \`.github/\` for automatic file-pattern matching.
5. **Copilot Chat** -- Reference any file directly in Copilot Chat for context-specific assistance.
`;
}

// ---------------------------------------------------------------------------
// Root copilot-instructions-root.md
// ---------------------------------------------------------------------------

function generateRootInstructions(allSkillsByCategory) {
    const sections = [];

    for (const category of CATEGORIES) {
        const skills = allSkillsByCategory[category] || [];
        if (skills.length === 0) continue;
        const title = toTitleCase(category);
        const list = skills.map(s => `  - ${s.displayName}`).join('\n');
        sections.push(`### ${title} (${skills.length})\n${list}`);
    }

    const totalCount = Object.values(allSkillsByCategory)
        .reduce((sum, arr) => sum + arr.length, 0);

    return `# Copilot Skills Library

> Auto-generated root index -- ${totalCount} skills across ${CATEGORIES.length} categories.

This repository provides GitHub Copilot custom instructions, agent skill definitions,
and path-specific instruction files converted from the master skill library.

## Quick Start

1. Browse by category below.
2. Copy a \`custom-instructions/{category}/{skill}.md\` file into your project's \`.github/copilot-instructions.md\`.
3. Or use \`agent-skills/\` and \`instructions/\` formats for advanced Copilot integration.

## Categories

${sections.join('\n\n')}

## Directory Layout

\`\`\`
copilot-skills/
  custom-instructions/{category}/{slug}.md      -- full instruction files
  agent-skills/{category}/{slug}/SKILL.md       -- agent skill definitions
  instructions/{category}-{slug}.instructions.md -- path-specific snippets
  {category}/README.md                          -- per-category index
  copilot-instructions-root.md                  -- this file
\`\`\`
`;
}

// ---------------------------------------------------------------------------
// Processing
// ---------------------------------------------------------------------------

function processSkill(category, slug, sourcePath) {
    const skillMdPath = path.join(sourcePath, 'SKILL.md');
    if (!fs.existsSync(skillMdPath)) {
        return { ok: false, error: `No SKILL.md found in ${category}/${slug}` };
    }

    let content;
    try {
        content = fs.readFileSync(skillMdPath, 'utf-8');
    } catch (err) {
        return { ok: false, error: `Cannot read ${skillMdPath}: ${err.message}` };
    }

    const { name, description, body } = parseSkillMd(content);
    const displayName = name ? toTitleCase(name) : toTitleCase(slug);

    // Paths
    const customDir = path.join(TARGET_BASE, 'custom-instructions', category);
    const agentDir  = path.join(TARGET_BASE, 'agent-skills', category, slug);
    const instrDir  = path.join(TARGET_BASE, 'instructions');

    ensureDir(customDir);
    ensureDir(agentDir);
    ensureDir(instrDir);

    try {
        // 1. Custom instruction
        const customContent = generateCustomInstruction(displayName, description, body, category, slug);
        fs.writeFileSync(path.join(customDir, `${slug}.md`), customContent, 'utf-8');

        // 2. Agent skill
        const agentContent = generateAgentSkill(name || slug, description, body);
        fs.writeFileSync(path.join(agentDir, 'SKILL.md'), agentContent, 'utf-8');

        // 3. Path-specific instruction
        const instrContent = generatePathInstruction(body, category);
        fs.writeFileSync(path.join(instrDir, `${category}-${slug}.instructions.md`), instrContent, 'utf-8');

        return { ok: true, displayName, slug };
    } catch (err) {
        return { ok: false, error: `Write failed for ${category}/${slug}: ${err.message}` };
    }
}

function processCategory(category) {
    const sourcePath = path.join(SOURCE_BASE, category);
    if (!fs.existsSync(sourcePath)) {
        return { skills: [], errors: [`Source directory missing: ${sourcePath}`] };
    }

    let items;
    try {
        items = fs.readdirSync(sourcePath, { withFileTypes: true });
    } catch (err) {
        return { skills: [], errors: [`Cannot read source directory ${sourcePath}: ${err.message}`] };
    }

    const skills = [];
    const errors = [];

    for (const item of items) {
        if (!item.isDirectory()) continue;

        const result = processSkill(category, item.name, path.join(sourcePath, item.name));
        if (result.ok) {
            skills.push({ displayName: result.displayName, slug: result.slug });
        } else {
            errors.push(result.error);
        }
    }

    return { skills, errors };
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

function main() {
    const startTime = Date.now();

    console.log('='.repeat(60));
    console.log('  GitHub Copilot Skill Converter');
    console.log('='.repeat(60));
    console.log();
    console.log(`  Source : ${SOURCE_BASE}`);
    console.log(`  Output : ${TARGET_BASE}`);
    console.log(`  Categories : ${CATEGORIES.join(', ')}`);
    console.log();

    ensureDir(TARGET_BASE);

    const allSkillsByCategory = {};
    let totalConverted = 0;
    let totalErrors = 0;
    const allErrors = [];

    for (const category of CATEGORIES) {
        const catStart = Date.now();
        process.stdout.write(`  [${category}] processing...`);

        const { skills, errors } = processCategory(category);

        allSkillsByCategory[category] = skills;
        totalConverted += skills.length;
        totalErrors += errors.length;
        allErrors.push(...errors.map(e => `[${category}] ${e}`));

        const catMs = Date.now() - catStart;
        // Clear the "processing..." and rewrite with results
        process.stdout.write(`\r  [${category}] ${skills.length} skills converted`);
        if (errors.length > 0) {
            process.stdout.write(` (${errors.length} errors)`);
        }
        console.log(` [${catMs}ms]`);
    }

    console.log();

    // -- Category README files --
    process.stdout.write('  Writing category READMEs...');
    for (const category of CATEGORIES) {
        const skills = allSkillsByCategory[category] || [];
        if (skills.length === 0) continue;
        const readmePath = path.join(TARGET_BASE, category);
        ensureDir(readmePath);
        const readmeContent = generateCategoryReadme(category, skills);
        fs.writeFileSync(path.join(readmePath, 'README.md'), readmeContent, 'utf-8');
    }
    console.log(' done');

    // -- Root instructions file --
    process.stdout.write('  Writing copilot-instructions-root.md...');
    const rootContent = generateRootInstructions(allSkillsByCategory);
    fs.writeFileSync(path.join(TARGET_BASE, 'copilot-instructions-root.md'), rootContent, 'utf-8');
    console.log(' done');

    // -- Summary --
    const elapsed = Date.now() - startTime;
    console.log();
    console.log('-'.repeat(60));
    console.log(`  Total skills converted : ${totalConverted}`);
    console.log(`  Output formats         : 3 (custom-instructions, agent-skills, instructions)`);
    console.log(`  Category READMEs       : ${CATEGORIES.length}`);
    console.log(`  Errors                 : ${totalErrors}`);
    console.log(`  Time                   : ${elapsed}ms`);
    console.log('-'.repeat(60));

    if (allErrors.length > 0) {
        console.log();
        console.log('  Errors:');
        for (const e of allErrors) {
            console.log(`    - ${e}`);
        }
    }

    console.log();
    console.log('  Conversion complete.');
    console.log();

    // Exit with non-zero if there were any category-level errors (missing dirs, etc.)
    // Individual missing SKILL.md files are warnings, not fatal.
    const fatalErrors = allErrors.filter(e => e.includes('Source directory missing'));
    if (fatalErrors.length > 0) {
        process.exit(1);
    }
}

main();
